File: C++/프로그래머스/3/43164. 여행경로/여행경로.cpp
#include <string>
#include <vector>
#include <map>
#include <algorithm>

using namespace std;

vector<string> vec;
map<string, vector<string>> graph;
map<string, vector<bool>> used;

bool dfs(string cur, vector<string> path, int totalTickets)
{
    if(path.size() == totalTickets)
    {
        vec = path;
        return true;
    }
    
    for(int i = 0; i < graph[cur].size(); i++)
    {
        if(!used[cur][i])
        {
            string temp = graph[cur][i];
            
            used[cur][i] = true;
            path.push_back(temp);
            
            if(dfs(temp, path, totalTickets)) return true;
            
            path.pop_back();
            used[cur][i] = false;
        }
    }
    
    return false;
}

vector<string> solution(vector<vector<string>> tickets)
{
    vector<string> answer;
    
    for(auto& t : tickets)
    {
        used[t[0]].push_back(false);
        graph[t[0]].push_back(t[1]);
        sort(graph[t[0]].begin(), graph[t[0]].end());
    }
    
    dfs("ICN", {"ICN"}, tickets.size() + 1);
    
    answer = vec;
              
    return answer;
}

File: Python3/프로그래머스/3/43164. 여행경로/여행경로.py
def dfs(cur, path, dic, used, total):
    if len(path) == total:
        return True
    if cur not in dic:
        return False
    
    for i in range(len(dic[cur])):
        if used[cur][i]:
            continue
            
        used[cur][i] = True
        nxt = dic[cur][i]
        path.append(nxt)
            
        if(dfs(nxt, path, dic, used, total)): return True
            
        used[cur][i] = False
        path.pop()
    
    return False

def solution(tickets):
    answer = ["ICN"]
    
    dic = {}
    used = {}
    for t in tickets:
        if t[0] not in dic.keys():
            dic[t[0]] = [t[1]]
            used[t[0]] = [False]
        else:
            dic[t[0]].append(t[1])
            used[t[0]].append(False)
        
            dic[t[0]].sort()
    
    dfs("ICN", answer, dic, used, len(tickets) + 1)
    
    return answer